pkg/foundation/csync/valuewatcher_test.go

```diff
@@ -0,0 +1,240 @@
+// Copyright © 2022 Meroxa, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package csync
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/matryer/is"
+	"go.uber.org/goleak"
+)
+
+func TestValueWatcher_GetEmptyValue(t *testing.T) {
+	is := is.New(t)
+
+	var h ValueWatcher[int]
+	got := h.Get()
+	is.Equal(0, got)
+}
+
+func TestValueWatcher_GetEmptyPtr(t *testing.T) {
+	is := is.New(t)
+
+	var h ValueWatcher[*int]
+	got := h.Get()
+	is.Equal(nil, got)
+}
+
+func TestValueWatcher_PutGetValue(t *testing.T) {
+	is := is.New(t)
+
+	var h ValueWatcher[int]
+	want := 123
+	h.Set(want)
+	got := h.Get()
+	is.Equal(want, got)
+}
+
+func TestValueWatcher_PutGetPtr(t *testing.T) {
+	is := is.New(t)
+
+	var h ValueWatcher[*int]
+	want := 123
+	h.Set(&want)
+	got := h.Get()
+	is.Equal(&want, got)
+}
+
+func TestValueWatcher_WatchSuccess(t *testing.T) {
+	goleak.VerifyNone(t)
+	is := is.New(t)
+
+	var h ValueWatcher[int]
+
+	putValue := make(chan int)
+	defer close(putValue)
+	go func() {
+		for val := range putValue {
+			h.Set(val)
+		}
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	i := 0
+	val, err := h.Watch(ctx, func(val int) bool {
+		i++
+		switch i {
+		case 1:
+			is.Equal(0, val) // expected first value to be 0
```

Comment:
Do you think there's an alternative? I don't see one, given that `ValueWatcher` might already contain the value that you are looking for, the zero value might be this value. There's a valid use case in the SDK where we use this behavior to determine if we should stop a destination - Conduit tells the destination to stop after it receives some specific position, and if no records were written that position is `nil`, we don't want to wait for `nil` to be explicitly set in `ValueWatcher` ([snippet](https://github.com/ConduitIO/conduit-connector-sdk/blob/345231192036248047cd5e99f230ca6fc92de4f3/destination.go#L221-L224)).

Reply:
Maybe you can use a flag to indicate that the value has been set? The value I expect to get from the `ValueWatcher` is a value someone else put into it, not a default one.

Reply:
Why though? As I explained above, there's already a use case for being notified about the current value, even if it wasn't explicitly set yet. Besides, isn't it simpler to have the same behavior that normal Go variables already have? Once you declare a variable it contains the zero value of that type, `ValueWatcher` behaves the same way. Why would we create a different behavior here?

Reply:
> Besides, isn't it simpler to have the same behavior that normal Go variables already have? Once you declare a variable it contains the zero value of that type, `ValueWatcher` behaves the same way. Why would we create a different behavior here?

I don't think we should actually change the initial value a `ValueWatcher` holds. My suggestion was related to the semantics of the `Watch` method, which I associate with changes (and not the current value, whatever it may be), but we can also call that a matter of wording. I definitely recognize the need to also include the current value. What I'd suggest in that regard is to change the comments related to watching, which say that you can wait for a value to be put into the ValueWatcher.